storage/innobase/trx/trx0purge.cc

```diff
@@ -554,16 +579,62 @@ dberr_t purge_sys_t::iterator::free_history() const
       ut_ad(rseg.is_persistent());
       log_free_check();
       rseg.latch.wr_lock(SRW_LOCK_CALL);
-      dberr_t err=
-        trx_purge_truncate_rseg_history(rseg, *this, !rseg.is_referenced() &&
-                                        purge_sys.sees(rseg.needs_purge));
+      dberr_t err= free_history_rseg(rseg);
       rseg.latch.wr_unlock();
       if (err)
         return err;
     }
   return DB_SUCCESS;
 }
 
+/** Start truncating an undo tablespace.
+@return undo tablespace that is chosen for truncation
+@retval nullptr  if undo tablespace truncation is not enabled or possible */
+fil_space_t *purge_sys_t::truncating_tablespace()
+{
+  ut_ad(this == &purge_sys);
+
+  fil_space_t *space= truncate.current;
+  if (space || srv_undo_tablespaces_active < 2 || !srv_undo_log_truncate)
+    return space;
+
+  const ulint threshold= ulint(srv_max_undo_log_size >> srv_page_size_shift);
+  for (ulint i= truncate.last, j= i;; )
```

Comment:
I am used to FORTRAN-style loop variable names `i`, `j`, `k`, so that I don’t need to use my imagination. These variables occur only a few times in the loop. One thing that we could do to make the loop body shorter would be to introduce an inline function for most of the loop body. Would that be acceptable?

Reply:
sure, it should do.